File: sdk/java/src/main/java/io/cordum/cap/Middleware.java
package io.cordum.cap;

import io.cordum.cap.agent.v1.BusPacket;
import io.cordum.cap.agent.v1.JobRequest;
import io.cordum.cap.agent.v1.JobResult;

import java.util.List;
import java.util.logging.Logger;

public final class Middleware {

    private Middleware() {}

    /** Context passed through the middleware chain. */
    public static final class Context {
        private final String traceId;
        private final String senderId;
        private final BusPacket packet;

        public Context(String traceId, String senderId, BusPacket packet) {
            this.traceId = traceId;
            this.senderId = senderId;
            this.packet = packet;
        }

        public String traceId() { return traceId; }
        public String senderId() { return senderId; }
        public BusPacket packet() { return packet; }
    }

    /** Processes a job request and returns a result. */
    @FunctionalInterface
    public interface Handler {
        JobResult handle(Context ctx, JobRequest request) throws Exception;
    }

    /** Wraps a handler, returning a new handler. */
    @FunctionalInterface
    public interface Interceptor {
        Handler apply(Handler next);
    }

    /**
     * Compose a chain of interceptors around a base handler.
     * Interceptors execute in FIFO order (first added = outermost wrapper).
     */
    public static Handler applyMiddleware(Handler base, List<Interceptor> chain) {
        Handler handler = base;
        for (int i = chain.size() - 1; i >= 0; i--) {
            handler = chain.get(i).apply(handler);
        }
        return handler;
    }

    /** Interceptor that logs job_id and duration for each request. */
    public static Interceptor loggingMiddleware(Logger logger) {
        return next -> (ctx, req) -> {
            long start = System.currentTimeMillis();
            JobResult result = next.handle(ctx, req);
            long elapsed = System.currentTimeMillis() - start;
            logger.info("[cap] job=" + req.getJobId() + " duration=" + elapsed + "ms");
            return result;
        };
    }
}

File: sdk/dotnet/src/Cordum.Cap/Middleware.cs
using System.Diagnostics;
using Cordum.Agent.V1;

namespace Cordum.Cap;

/// <summary>
/// Context passed through the middleware chain.
/// </summary>
public sealed class Context
{
    public required string TraceId { get; init; }
    public required string SenderId { get; init; }
    public BusPacket? Packet { get; init; }
}

/// <summary>
/// Synchronous job handler delegate.
/// </summary>
public delegate JobResult HandlerDelegate(Context ctx, JobRequest request);

/// <summary>
/// Middleware wraps a handler, returning a new handler.
/// </summary>
public delegate HandlerDelegate MiddlewareDelegate(HandlerDelegate next);

/// <summary>
/// Middleware composition utilities.
/// </summary>
public static class MiddlewareChain
{
    /// <summary>
    /// Compose a chain of middlewares around a base handler.
    /// Middlewares execute in FIFO order (first added = outermost wrapper).
    /// </summary>
    public static HandlerDelegate ApplyMiddleware(HandlerDelegate handler, IReadOnlyList<MiddlewareDelegate> chain)
    {
        var result = handler;
        for (var i = chain.Count - 1; i >= 0; i--)
        {
            result = chain[i](result);
        }
        return result;
    }

    /// <summary>
    /// Middleware that logs job_id and duration for each request.
    /// </summary>
    public static MiddlewareDelegate LoggingMiddleware()
    {
        return next => (ctx, req) =>
        {
            var sw = Stopwatch.StartNew();
            var result = next(ctx, req);
            sw.Stop();
            Console.WriteLine($"[cap] job={req.JobId} duration={sw.ElapsedMilliseconds}ms");
            return result;
        };
    }
}
